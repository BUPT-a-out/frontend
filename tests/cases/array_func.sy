const int g = 14;
const int N = 10000;

int data[N] = {0, 1, 2, 3, 4};
int data2[N][5] = {{1}, {2, 3}};

int sum(int a[], int arr[], int arr2[][5]) {
    return a[0] + arr[0] + arr[1] + arr2[0][0] + arr2[0][1];
}

int main() {
    int a = 5;
    int b[10] = {4, 3, 2, 1};
    int c[5][5] = {{1, 2}, {3, 4}};
    return a + data[1] + b[3] + g + sum(b, c[0], data2);
}