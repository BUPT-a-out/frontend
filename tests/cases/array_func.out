Parsing completed successfully.

--- Abstract Syntax Tree ---
+-- ROOT
|   +-- CONST_VAR_DEF: g (sym: g, id: 26)
|   |   +-- CONST (int value: 14)
|   +-- CONST_VAR_DEF: N (sym: N, id: 27)
|   |   +-- CONST (int value: 10000)
|   +-- ARRAY_DEF: data (sym: data, id: 28)
|   |   +-- LIST: ArrayIniter
|   |   |   +-- CONST (int value: 0)
|   |   |   +-- CONST (int value: 1)
|   |   |   +-- CONST (int value: 2)
|   |   |   +-- CONST (int value: 3)
|   |   |   +-- CONST (int value: 4)
|   +-- ARRAY_DEF: data2 (sym: data2, id: 29)
|   |   +-- LIST: ArrayIniter
|   |   |   +-- LIST: ArrayIniter
|   |   |   |   +-- CONST (int value: 1)
|   |   |   +-- LIST: ArrayIniter
|   |   |   |   +-- CONST (int value: 2)
|   |   |   |   +-- CONST (int value: 3)
|   +-- FUNC_DEF: sum (sym: sum, id: 30)
|   |   +-- LIST: FParams
|   |   |   +-- ARRAY_DEF: a (sym: a, id: 31)
|   |   |   +-- ARRAY_DEF: arr (sym: arr, id: 32)
|   |   |   +-- ARRAY_DEF: arr2 (sym: arr2, id: 33)
|   |   +-- LIST: Block
|   |   |   +-- RETURN_STMT: sum (sym: sum, id: 30)
|   |   |   |   +-- BINARY_OP: +
|   |   |   |   |   +-- BINARY_OP: +
|   |   |   |   |   |   +-- BINARY_OP: +
|   |   |   |   |   |   |   +-- BINARY_OP: +
|   |   |   |   |   |   |   |   +-- ARRAY_ACCESS: Dims (sym: a, id: 31)
|   |   |   |   |   |   |   |   |   +-- CONST (int value: 0)
|   |   |   |   |   |   |   |   +-- ARRAY_ACCESS: Dims (sym: arr, id: 32)
|   |   |   |   |   |   |   |   |   +-- CONST (int value: 0)
|   |   |   |   |   |   |   +-- ARRAY_ACCESS: Dims (sym: arr, id: 32)
|   |   |   |   |   |   |   |   +-- CONST (int value: 1)
|   |   |   |   |   |   +-- ARRAY_ACCESS: Dims (sym: arr2, id: 33)
|   |   |   |   |   |   |   +-- CONST (int value: 0)
|   |   |   |   |   |   |   +-- CONST (int value: 0)
|   |   |   |   |   +-- ARRAY_ACCESS: Dims (sym: arr2, id: 33)
|   |   |   |   |   |   +-- CONST (int value: 0)
|   |   |   |   |   |   +-- CONST (int value: 1)
|   +-- FUNC_DEF: main (sym: main, id: 34)
|   |   +-- LIST: FParams
|   |   +-- LIST: Block
|   |   |   +-- VAR_DEF: a (sym: a, id: 35)
|   |   |   |   +-- CONST (int value: 5)
|   |   |   +-- ARRAY_DEF: b (sym: b, id: 36)
|   |   |   |   +-- LIST: ArrayIniter
|   |   |   |   |   +-- CONST (int value: 4)
|   |   |   |   |   +-- CONST (int value: 3)
|   |   |   |   |   +-- CONST (int value: 2)
|   |   |   |   |   +-- CONST (int value: 1)
|   |   |   +-- ARRAY_DEF: c (sym: c, id: 37)
|   |   |   |   +-- LIST: ArrayIniter
|   |   |   |   |   +-- LIST: ArrayIniter
|   |   |   |   |   |   +-- CONST (int value: 1)
|   |   |   |   |   |   +-- CONST (int value: 2)
|   |   |   |   |   +-- LIST: ArrayIniter
|   |   |   |   |   |   +-- CONST (int value: 3)
|   |   |   |   |   |   +-- CONST (int value: 4)
|   |   |   +-- RETURN_STMT: main (sym: main, id: 34)
|   |   |   |   +-- BINARY_OP: +
|   |   |   |   |   +-- BINARY_OP: +
|   |   |   |   |   |   +-- BINARY_OP: +
|   |   |   |   |   |   |   +-- BINARY_OP: +
|   |   |   |   |   |   |   |   +-- VAR (sym: a, id: 35)
|   |   |   |   |   |   |   |   +-- ARRAY_ACCESS: Dims (sym: data, id: 28)
|   |   |   |   |   |   |   |   |   +-- CONST (int value: 1)
|   |   |   |   |   |   |   +-- ARRAY_ACCESS: Dims (sym: b, id: 36)
|   |   |   |   |   |   |   |   +-- CONST (int value: 3)
|   |   |   |   |   |   +-- CONST (int value: 14)
|   |   |   |   |   +-- FUNC_CALL: sum (sym: sum, id: 30)
|   |   |   |   |   |   +-- ARRAY (sym: b, id: 36)
|   |   |   |   |   |   +-- ARRAY_ACCESS: Dims (sym: c, id: 37)
|   |   |   |   |   |   |   +-- CONST (int value: 0)
|   |   |   |   |   |   +-- ARRAY (sym: data2, id: 29)

--- Permanent Symbol Table ---
ID    Name                 Type            Data Type  Function             Shape     
-------------------------------------------------------------------------------------
0     getint               function        int        N/A                  N/A       
1     getch                function        int        N/A                  N/A       
2     getfloat             function        float      N/A                  N/A       
3     getarray             function        int        N/A                  N/A       
4     array                array           int        getarray             0
5     getfarray            function        int        N/A                  N/A       
6     array                array           float      getfarray            0
7     putint               function        void       N/A                  N/A       
8     value                var             int        putint               N/A       
9     putch                function        void       N/A                  N/A       
10    value                var             int        putch                N/A       
11    putfloat             function        void       N/A                  N/A       
12    value                var             float      putfloat             N/A       
13    putarray             function        void       N/A                  N/A       
14    len                  var             int        putarray             N/A       
15    array                array           int        putarray             0
16    putfarray            function        void       N/A                  N/A       
17    len                  var             int        putfarray            N/A       
18    array                array           float      putfarray            0
19    putf                 function        void       N/A                  N/A       
20    format_string        var             char       putf                 N/A       
21    value                var             int        putf                 N/A       
22    starttime            function        void       N/A                  N/A       
23    line                 var             int        starttime            N/A       
24    stoptime             function        void       N/A                  N/A       
25    line                 var             int        stoptime             N/A       
26    g                    const var       int        N/A                  N/A       
27    N                    const var       int        N/A                  N/A       
28    data                 array           int        N/A                  10000
29    data2                array           int        N/A                  10000 5
30    sum                  function        int        N/A                  N/A       
31    a                    array           int        sum                  0
32    arr                  array           int        sum                  0
33    arr2                 array           int        sum                  0 5
34    main                 function        int        N/A                  N/A       
35    a                    var             int        main                 N/A       
36    b                    array           int        main                 10
37    c                    array           int        main                 5 5
-------------------------------------------------------------------------------------

--- Generated IR ---
; ModuleID = 'main'

@g.26 = internal global i32 14
@N.27 = internal global i32 10000
@data.28 = external global [10000 x i32] [i32 0, i32 1, i32 2, i32 3, i32 4]
@data2.29 = external global [10000 x [5 x i32]] [[5 x i32] [i32 1], [5 x i32] [i32 2, i32 3]]

define i32 @sum(i32* %param.a.31, i32* %param.arr.32, [5 x i32]* %param.arr2.33) {
sum.entry:
  %0 = getelementptr i32*, i32* %param.a.31, i32 0
  %1 = load i32, i32* %0
  %2 = getelementptr i32*, i32* %param.arr.32, i32 0
  %3 = load i32, i32* %2
  %add.4 = add i32 %1, %3
  %5 = getelementptr i32*, i32* %param.arr.32, i32 1
  %6 = load i32, i32* %5
  %add.7 = add i32 %add.4, %6
  %8 = getelementptr [5 x i32]*, [5 x i32]* %param.arr2.33, i32 0, i32 0
  %9 = load i32, i32* %8
  %add.10 = add i32 %add.7, %9
  %11 = getelementptr [5 x i32]*, [5 x i32]* %param.arr2.33, i32 0, i32 1
  %12 = load i32, i32* %11
  %add.13 = add i32 %add.10, %12
  ret i32 %add.13
}

define i32 @main() {
main.entry:
  %a.35 = alloca i32
  %b.36 = alloca [10 x i32]
  %c.37 = alloca [5 x [5 x i32]]
  store i32 5, i32* %a.35
  %b.36.initer = alloca i32
  store i32 0, i32* %b.36.initer
  br label %b.36.initer.while.cond
b.36.initer.while.cond:
  %18 = load i32, i32* %b.36.initer
  %lt.19 = icmp slt i32 %18, 10
  br i1 %lt.19, label %b.36.initer.while.loop, label %b.36.initer.while.merge
b.36.initer.while.loop:
  %14 = load i32, i32* %b.36.initer
  %15 = getelementptr [10 x i32], [10 x i32]* %b.36, i32 %14
  store i32 0, i32* %15
  %16 = load i32, i32* %b.36.initer
  %17 = add i32 %16, 1
  store i32 %17, i32* %b.36.initer
  br label %b.36.initer.while.cond
b.36.initer.while.merge:
  %20 = getelementptr [10 x i32], [10 x i32]* %b.36, i32 0
  store i32 4, i32* %20
  %21 = getelementptr [10 x i32], [10 x i32]* %b.36, i32 1
  store i32 3, i32* %21
  %22 = getelementptr [10 x i32], [10 x i32]* %b.36, i32 2
  store i32 2, i32* %22
  %23 = getelementptr [10 x i32], [10 x i32]* %b.36, i32 3
  store i32 1, i32* %23
  %c.37.initer = alloca i32
  store i32 0, i32* %c.37.initer
  br label %c.37.initer.while.cond
c.37.initer.while.cond:
  %28 = load i32, i32* %c.37.initer
  %lt.29 = icmp slt i32 %28, 25
  br i1 %lt.29, label %c.37.initer.while.loop, label %c.37.initer.while.merge
c.37.initer.while.loop:
  %24 = load i32, i32* %c.37.initer
  %25 = getelementptr [25 x i32], [5 x [5 x i32]]* %c.37, i32 %24
  store i32 0, i32* %25
  %26 = load i32, i32* %c.37.initer
  %27 = add i32 %26, 1
  store i32 %27, i32* %c.37.initer
  br label %c.37.initer.while.cond
c.37.initer.while.merge:
  %30 = getelementptr [25 x i32], [5 x [5 x i32]]* %c.37, i32 0
  store i32 1, i32* %30
  %31 = getelementptr [25 x i32], [5 x [5 x i32]]* %c.37, i32 1
  store i32 2, i32* %31
  %32 = getelementptr [25 x i32], [5 x [5 x i32]]* %c.37, i32 5
  store i32 3, i32* %32
  %33 = getelementptr [25 x i32], [5 x [5 x i32]]* %c.37, i32 6
  store i32 4, i32* %33
  %34 = load i32, i32* %a.35
  %35 = getelementptr [10000 x i32], [10000 x i32]* @data.28, i32 1
  %36 = load i32, i32* %35
  %add.37 = add i32 %34, %36
  %38 = getelementptr [10 x i32], [10 x i32]* %b.36, i32 3
  %39 = load i32, i32* %38
  %add.40 = add i32 %add.37, %39
  %add.41 = add i32 %add.40, 14
  %42 = getelementptr [5 x [5 x i32]], [5 x [5 x i32]]* %c.37, i32 0
  %43 = call i32 @sum([10 x i32]* %b.36, [5 x i32]* %42, [10000 x [5 x i32]]* @data2.29)
  %add.44 = add i32 %add.41, %43
  ret i32 %add.44
}


